### PR TITLE
fix: valid id token

### PR DESCRIPTION
**Pull request by danieljhegeman**
- #3945 

### Reviewers
**Functional:** 
@tihuan @Bento007 
**Readability:** 

---

## Changes
- use a real key pair (this may actually not be necessary but it is nice to have in place for future-proofing)
- use nonce claim, pulled from header in initial authorize request (required)
- use client id as audience claim value (required)

## QA steps (optional)
- rebuild oidc container `docker-compose build oidc`
- run local dev mode
- click 'Log In' in top right; should see `Fake User` appear as logged-in user.

### CODE REVIEW

File: frontend/next.config.js [hunk 1]
```diff
@@ -41,7 +41,7 @@ const defaultSecureHeaders = {
       scriptSrc: isProdBuild ? SCRIPT_SRC : [...SCRIPT_SRC, "'unsafe-eval'"],
       styleSrc: ["'self'", "'unsafe-inline'", "https://fonts.googleapis.com"],
       upgradeInsecureRequests: true,
-      workerSrc: true,
+      workerSrc: ["blob:"],
```
Comment (danieljhegeman) [inline frontend/next.config.js:44]: from https://github.com/chanzuckerberg/single-cell-data-portal/pull/3952

File: frontend/src/pages/_app.tsx [hunk 2]
```diff
@@ -66,6 +66,7 @@ function App({ Component, pageProps }: AppPropsWithLayout): JSX.Element {
         clientId={AUTH0_CLIENT_ID}
         domain={AUTH0_DOMAIN}
         redirectUri={redirectUri}
+        useRefreshTokens={true}
```
Comment (danieljhegeman) [inline frontend/src/pages/_app.tsx:69]: from https://github.com/chanzuckerberg/single-cell-data-portal/pull/3952

File: oauth/mock_auth0_pkce.py [hunk 3]
```diff
@@ -101,12 +114,48 @@ def __init__(self, additional_scope=None, token_duration=0):
...
+    issued_at = now
+    expires_at = now + token_duration
+    headers = dict(alg="RS256")
+    token_claims.update(
+        exp=expires_at,
+        iat=issued_at,
+        # For some reason, Auth0 requires the client id here for the audience value even though the Auth0Provider takes
+        # distinct prop values 'audience' and 'clientId'. Without this configuration, the Auth0 React package rejects
+        # the JWT.
+        aud="local-client-id",
```
Comment (danieljhegeman) [inline oauth/mock_auth0_pkce.py:128]: this has to correspond to `clientId` prop value from [here](https://github.com/chanzuckerberg/single-cell-data-portal/blob/17e490d8bd1b44074efabe1709db469c61c57331/frontend/src/pages/_app.tsx#L66) and notably _not_ the `audience` prop value immediately above it. 🤬🤦‍♂️

File: oauth/mock_auth0_pkce.py [hunk 4]
```diff
@@ -101,12 +114,48 @@ def __init__(self, additional_scope=None, token_duration=0):
...
+        scope=additional_scope,
+        iss="https://oidc.corporanet.local/",
```
Comment (danieljhegeman) [inline oauth/mock_auth0_pkce.py:130]: permits verification of signature

File: oauth/mock_auth0_pkce.py (hunk 3 above)
Comment (tihuan) [reply oauth/mock_auth0_pkce.py:128]: Oh wow that's crazy! Thanks for adding the comment for this gotcha 😵‍💫

**Comment by tihuan** [comment]
I'll defer to @Bento007 for approval 🙏 Thanks all!

**Review by Bento007 — COMMENTED** [review]
- Add some logging
- Does this docker container replacer the old OIDC container?
- I got a 500 when using the portal API with cxguser cookie on the GET /userinfo endpoint. Is this expected?

**Review by Bento007 — APPROVED** [review]

**Pull request merged**